specs/ledger/hs/src/Ledger/Update.hs

```diff
@@ -539,3 +542,244 @@ instance STS UPEND where
+  wrapFailed = UPVOTEFailure
+
+data UPIEND
+
+instance STS UPIEND where
+  type Environment UPIEND = UPIEnv
+  type State UPIEND = UPIState
+  type Signal UPIEND = (ProtVer, Core.VKey)
+  data PredicateFailure UPIEND
+    = UPENDFailure (PredicateFailure UPEND)
+    deriving (Eq, Show)
+
+  initialRules = []
+  transitionRules =
+    [ do
+        TRC ( (_ec, sn, _dms)
+            , ( ep
+              , (pv, pps)
+              , fads
+              , avs
+              , rpus
+              , raus
+              , cps
+              , vts
+              , bvs
+              , pws)
+            , (bv,vk)) <- judgmentContext
+        let k = pps ^. chainStability
+            u = pps ^. upTtl
+        (fads', bvs') <- trans @UPEND $ TRC ((k, sn, (pv, pps), cps, rpus), (fads, bvs), (bv,vk))
+        let pidskeep = dom (Map.filter (> (sn `Core.minusSlot` u)) pws) `Set.union` dom cps
+            rpus' = pidskeep ◃ rpus
+            vskeep = Set.fromList . fmap fst $ Map.elems rpus'
+        return ( ep
+                , (pv, pps)
+                , fads'
+                , avs
+                , rpus'
+                , pidskeep ◃ raus
+                , cps
+                , pidskeep ◃ vts
+                , vskeep ◃ bvs'
+                , pidskeep ◃ pws
+                )
+
+    ]
+
+instance Embed UPEND UPIEND where
+  wrapFailed = UPENDFailure
+
+data PVBUMP
+
+instance STS PVBUMP where
+  type Environment PVBUMP =
+    ( Core.SlotCount
+    , Core.Slot
+    )
+  type State PVBUMP =
+    ( Core.Epoch
+    , (ProtVer, PParams)
+    , [(Core.Slot, (ProtVer, PParams))]
+    )
+  type Signal PVBUMP = Core.Epoch
+  data PredicateFailure PVBUMP
+    = NewEpoch
+    | OldEpoch
+    deriving (Eq, Show)
+
+  initialRules = []
+  transitionRules =
+    [ do
+        TRC (_, (ep, (pv, pps), fads), en) <- judgmentContext
+        en <= ep ?! NewEpoch
+        return (ep, (pv, pps), fads)
+    , do
+        TRC ((k, sn), (ep, (pv, pps), fads), en) <- judgmentContext
+        ep < en ?! OldEpoch
+        return $ case (partition (\(s, _) -> s < sn `Core.minusSlot` (2*k)) fads) of
+          ([], _) -> (ep, (pv, pps), fads)
+          ((_, (pvc, ppsc)):_, rest) -> (en, (pvc, ppsc), rest)
+    ]
+
+data UPIEC
+
+instance STS UPIEC where
+  type Environment UPIEC = UPIEnv
+  type State UPIEC = UPIState
+  type Signal UPIEC = Core.Epoch
+  data PredicateFailure UPIEC
+    = PVBUMPFailure (PredicateFailure PVBUMP)
+    deriving (Eq, Show)
+
+  initialRules = []
+  transitionRules =
+    [ do
+        TRC ( (_ec, sn, _dms)
+            , ( ep
+              , (pv, pps)
+              , fads
+              , avs
+              , rpus
+              , raus
+              , cps
+              , vts
+              , bvs
+              , pws)
+            , en) <- judgmentContext
+        let k = pps ^. chainStability
+        (e', (pv', pps'), fads') <- trans @PVBUMP $ TRC ((k, sn), (ep, (pv, pps), fads), en)
+        let pidskeep = undefined
```

Comment:
Is the idea that we implement this later, or you just forgot it?

Reply:
Most likely the latter...